### PR TITLE
DEVX-1662: Update clickstream to sessionize data

### PR DESCRIPTION
**Pull request by bbejeck**
Updates the clickstream demo to use sessionized data.  This is accomplished by pausing/resuming the clickstream datagen connectors.

Also, ksqlDB populates the `EVENT_TS` field with the `WINDOWSTART` column vs. using an injected field from connect.  The expected clickstream dashboard is updated as well.

### CODE REVIEW

File: clickstream/start.sh [hunk 1]
```diff
@@ -43,3 +43,22 @@ Navigate to:
...
+
+echo -e "\nSleeping 30 seconds then starting clickstream sessionizing\n"
+sleep 30
+
+for i in {1..100}; do
+
+ 
+ docker-compose exec ksqldb-server curl -X PUT localhost:8083/connectors/DATAGEN_CLICKSTREAM/pause
+
+ echo -e "\nPaused the clickstream generate for 35 seconds\n"
```
Comment (ybyzek) [inline clickstream/start.sh]:
```suggestion
 echo -e "\nPaused the clickstream datagen for 35 seconds\n"
```

File: clickstream/start.sh [hunk 2]
```diff
@@ -43,3 +43,22 @@ Navigate to:
...
+ sleep 35
+
+
+ docker-compose exec ksqldb-server curl -X PUT localhost:8083/connectors/DATAGEN_CLICKSTREAM/resume
+
+ echo -e "\Resumed the clickstream generate for 90 seconds. Then the cycle will repeat\n"
```
Comment (ybyzek) [inline clickstream/start.sh]:
```suggestion
 echo -e "\Resumed the clickstream datagen for 90 seconds. Then the cycle will repeat\n"
```

**Review by ybyzek — CHANGES_REQUESTED** [review]
Great PR @bbejeck .

Additional comments:

1. Please update the image https://github.com/confluentinc/examples/blob/5.5.x/clickstream/images/clickstream-dashboard.png

2. The docs page (source: https://github.com/confluentinc/examples/blob/5.5.x/clickstream/docs/index.rst ... rendered: https://docs.confluent.io/current/ksql/docs/tutorials/clickstream-docker.html) has users walk through the demo one step at a time, users don't have to run the start.sh script.  So the part that does the pause/resume should be (a) extracted into a separate script and let users run the separate script or (b) explained how to run manually by users in the docs.  Either way, a docs update is warranted.

3. Can you add a comment into the script itself to explain the purpose of toggling the connector?

**Comment by bbejeck** [comment]
@ybyzek 
For 1, the clickstream image has been updated with this PR
https://github.com/confluentinc/examples/blob/DEVX-1662_update_clickstream_to_sessionize_data/clickstream/images/clickstream-dashboard.png. 

I didn't realize Github only shows the diff In size for binary files, and you can't view them; otherwise, I would have put the link in the commit comments, sorry about the confusion.

Ack on 2 and 3 from comments above.

**Comment by bbejeck** [comment]
@ybyzek comments addressed

File: clickstream/sessionize-data.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+# To simulate these user sessions, the script pauses the DATAGEN_CLICKSTREAM connector 
+# every 90 seconds for a 35 second period of inactivity.  By stopping the DATAGEN_CLICKSTREAM connector 
+# for more than 30 seconds, you are guaranteed to see user sessions.
+
+# Session windows are because they are driven by user behavior and
+# other window implementations are driven only by time.
+
+
+echo "Starting to run the session data, I'll pause for 90 seconds to allow more data to generate"
```
Comment (ybyzek) [inline clickstream/sessionize-data.sh]:
```suggestion
echo "Running for 90 seconds to allow clickstream data to generate"
```

File: clickstream/sessionize-data.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+sleep 90
+for i in {1..100}; do
+
+ 
+ docker-compose exec ksqldb-server curl -X PUT localhost:8083/connectors/DATAGEN_CLICKSTREAM/pause
+
+ echo "Paused the clickstream datagen for 35 seconds"
```
Comment (ybyzek) [inline clickstream/sessionize-data.sh]:
```suggestion
 echo "Pausing the clickstream datagen for 35 seconds"
```

File: clickstream/sessionize-data.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ sleep 35
+
+
+ docker-compose exec ksqldb-server curl -X PUT localhost:8083/connectors/DATAGEN_CLICKSTREAM/resume
+
+ echo "Resumed the clickstream datagen for 90 seconds. Then the cycle will repeat"
```
Comment (ybyzek) [inline clickstream/sessionize-data.sh]:
```suggestion
 echo "Resuming the clickstream datagen for 90 seconds. Then the cycle will repeat"
```

**Review by ybyzek — COMMENTED** [review]
This looks really good @bbejeck, excellent description.

I noticed two other things that we should probably fix (and feel free to assign this to me if you prefer I touch this up):

1. Sequencing in the docs/index.rst -- the screenshot `docs/images/grafana-success.png` comes BEFORE the section on sessionizing the data.  The "success" screenshot should come after the data starts getting sessionized.

2. `images/clickstream-dashboard.png` is an unnecessary duplicate of `docs/images/grafana-success.png`.  Perhaps we should do:

```
cp images/clickstream-dashboard.png docs/images/grafana-success.png
rm -fr images
... # and then replace wherever images/clickstream-dashboard.png is referenced (which is nowhere?)
```

**Comment by bbejeck** [comment]
Thanks for the review @ybyzek.

>Sequencing in the docs/index.rst -- the screenshot docs/images/grafana-success.png comes BEFORE the section on sessionizing the data. The "success" screenshot should come after the data starts getting sessionized.

I had questions myself about the order of things.  In this update, I added the `sessionizing` section entirely before the `grafana` section.  I think this should be ok, but let me know if want something more along the lines of

1. Instructions for Grafana
2. Instructions for session data
3. Then show grafana success image.

**Comment by bbejeck** [comment]
I remembered the sessionizing is a blocking script so I updated to the `index.rst` page to have the following order:

1. Instructions for Grafana
2. Instructions for session data
3. Then show grafana success image

@ybyzek WDYT?

**Review by ybyzek — APPROVED** [review]
@bbejeck very nice work!  LGTM

**Comment by bbejeck** [comment]
merged #547 into 5.5.x

**Comment by bbejeck** [comment]
merged into master with pint

**Pull request merged**